File: csharp/binary-search/BinarySearch.cs
using System;

public static class BinarySearch
{
    public static int Find(Span<int> array, int value, int offset = 0)
    {
        if (array.Length == 0)
            return -1;
        
        var middleIndex = array.Length / 2;
        var middleValue = array[middleIndex];
            
        if (middleValue == value)
            return offset + middleIndex;

        return middleValue > value
            ? Find(array.Slice(0, middleIndex), value, offset)
            : Find(array.Slice(middleIndex + 1), value, offset + middleIndex + 1);
    }
}

File: haskell/binary-search/src/BinarySearch.hs
module BinarySearch (find) where

import Data.Array

find :: Ord a => Array Int a -> a -> Maybe Int
find array x = findBetweenIndices array x 0 (length array - 1)

findBetweenIndices :: Ord a => Array Int a -> a -> Int -> Int-> Maybe Int
findBetweenIndices array x left right
    | left > right = Nothing
    | middle_value < x = findBetweenIndices array x (middle_index + 1) right
    | middle_value > x = findBetweenIndices array x left (middle_index - 1)
    | otherwise = Just middle_index
    where
        middle_index = (left + right) `quot` 2
        middle_value = array ! middle_index
